src/mongocxx/instance.cpp

```diff
@@ -94,31 +96,25 @@ instance::instance() : instance(nullptr) {
 }
 
 instance::instance(std::unique_ptr<logger> logger) {
-    std::lock_guard<std::recursive_mutex> lock(instance_mutex);
-    if (current_instance) {
+    instance* expected = nullptr;
+    if (!current_instance.compare_exchange_strong(expected, this))
         throw logic_error{error_code::k_instance_already_exists};
-    }
     _impl = stdx::make_unique<impl>(std::move(logger));
-    current_instance = this;
 }
 
 instance::instance(instance &&) noexcept = default;
 instance &instance::operator=(instance &&) noexcept = default;
 
 instance::~instance() {
-    std::lock_guard<std::recursive_mutex> lock(instance_mutex);
-    if (current_instance != this) std::abort();
     _impl.reset();
-    current_instance = nullptr;
+    current_instance.store(nullptr);
```

Comment:
I wonder if you can handle this with a little state machine.

I.e.  Create a synthetic instance* that implies the shutting down state and make instance::current spin on it if current_instance.load() returns that specific value.

That way you can:

1. Set the current_instance over to the shutting down state (which forces a spin)
2. Run the libmongoc cleanup code
3. Set the current_instance to null
4. Allow the caller of instance::current in and allow it to create a new instance (which re-initializes)